core/net/ip/tcpip.c

```diff
@@ -526,222 +522,268 @@ tcpip_input(void)
+void
+tcpip_ipv6_output(void)
+{
+  uip_ipaddr_t ipaddr;
+  uip_ds6_nbr_t *nbr = NULL;
+  const uip_lladdr_t *linkaddr;
+  uip_ipaddr_t *nexthop;
 
-        if(annotate_has_last) {
-          printf("#L %u 0; red\n", annotate_last);
-        }
-        printf("#L %u 1; red\n", nexthop->u8[sizeof(uip_ipaddr_t) - 1]);
-        annotate_last = nexthop->u8[sizeof(uip_ipaddr_t) - 1];
-        annotate_has_last = 1;
-      }
-#endif /* TCPIP_CONF_ANNOTATE_TRANSMISSIONS */
-    }
+  if(uip_len == 0) {
+    return;
+  }
 
-    /* End of next hop determination */
+  if(uip_len > UIP_LINK_MTU) {
+    UIP_LOG("tcpip_ipv6_output: Packet too big");
+    goto exit;
+  }
+
+  if(uip_is_addr_unspecified(&UIP_IP_BUF->destipaddr)){
+    UIP_LOG("tcpip_ipv6_output: Destination address unspecified");
+    goto exit;
+  }
+
+  if(uip_is_addr_mcast(&UIP_IP_BUF->destipaddr)) {
+    linkaddr = NULL;
+    goto send_packet;
+  }
+
+  if((nexthop = get_nexthop(&ipaddr)) == NULL) {
+    goto exit;
+  }
+  annotate_transmission(nexthop);
 
 #if UIP_CONF_IPV6_RPL
-    if(!rpl_finalize_header(nexthop)) {
-      uip_clear_buf();
-      return;
-    }
+  if(!rpl_finalize_header(nexthop)) {
+    goto exit;
+  }
 #endif /* UIP_CONF_IPV6_RPL */
-    nbr = uip_ds6_nbr_lookup(nexthop);
-    if(nbr == NULL) {
-#if UIP_ND6_SEND_NA
-      if((nbr = uip_ds6_nbr_add(nexthop, NULL, 0, NBR_INCOMPLETE, NBR_TABLE_REASON_IPV6_ND, NULL)) == NULL) {
-        uip_clear_buf();
-        PRINTF("tcpip_ipv6_output: failed to add neighbor to cache\n");
-        return;
-      } else {
-#if UIP_CONF_IPV6_QUEUE_PKT
-        /* Copy outgoing pkt in the queuing buffer for later transmit. */
-        if(uip_packetqueue_alloc(&nbr->packethandle, UIP_DS6_NBR_PACKET_LIFETIME) != NULL) {
-          memcpy(uip_packetqueue_buf(&nbr->packethandle), UIP_IP_BUF, uip_len);
-          uip_packetqueue_set_buflen(&nbr->packethandle, uip_len);
-        }
-#endif
-        /* RFC4861, 7.2.2:
-         * "If the source address of the packet prompting the solicitation is the
-         * same as one of the addresses assigned to the outgoing interface, that
-         * address SHOULD be placed in the IP Source Address of the outgoing
-         * solicitation.  Otherwise, any one of the addresses assigned to the
-         * interface should be used."*/
-        if(uip_ds6_is_my_addr(&UIP_IP_BUF->srcipaddr)){
-          uip_nd6_ns_output(&UIP_IP_BUF->srcipaddr, NULL, &nbr->ipaddr);
-        } else {
-          uip_nd6_ns_output(NULL, NULL, &nbr->ipaddr);
-        }
 
-        stimer_set(&nbr->sendns, uip_ds6_if.retrans_timer / 1000);
-        nbr->nscount = 1;
-        /* Send the first NS try from here (multicast destination IP address). */
-      }
-#else /* UIP_ND6_SEND_NA */
-      PRINTF("tcpip_ipv6_output: neighbor not in cache\n");
-      uip_len = 0;
-      return;  
-#endif /* UIP_ND6_SEND_NA */
+  nbr = uip_ds6_nbr_lookup(nexthop);
+  if(nbr == NULL) {
+    if(send_nd6_ns(nexthop)) {
```

Comment:
This way you might even be able to refactor this with the `#if UIP_ND6_SEND_NA` at line 755